hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java

```diff
@@ -503,35 +489,18 @@ public CacheConfig findCacheConfig(String simpleName) {
         try {
             // Set name explicitly, because found config might have a wildcard name.
             CacheConfig cacheConfig = new CacheConfig(cacheSimpleConfig).setName(simpleName);
-            setTenantControl(cacheConfig);
             return cacheConfig;
         } catch (Exception e) {
             throw new CacheException(e);
         }
     }
 
-    @Override
-    public void setTenantControl(CacheConfig cacheConfig) {
-        if (!NOOP_TENANT_CONTROL.equals(getTenantControl(cacheConfig))) {
-            // a tenant control has already been explicitly set for the cache config
-            return;
-        }
-        // associate cache config with the current thread's tenant
-        // and add hook so when the tenant is destroyed, so is the cache config
-        TenantControlFactory tenantControlFactory = null;
-        try {
-            tenantControlFactory = ServiceLoader.load(TenantControlFactory.class,
-                    TENANT_CONTROL_FACTORY, nodeEngine.getConfigClassLoader());
-        } catch (Exception e) {
-            if (logger.isFinestEnabled()) {
-                logger.finest("Could not load service provider for TenantControl", e);
-            }
-        }
-        if (tenantControlFactory == null) {
-            tenantControlFactory = NOOP_TENANT_CONTROL_FACTORY;
-        }
-        CacheConfigAccessor.setTenantControl(cacheConfig, tenantControlFactory.saveCurrentTenant(
-                new CacheDestroyEventContext(cacheConfig.getName())));
+    public void reSerializeCacheConfig(CacheConfig cacheConfig) {
+        CompletableFuture<CacheConfig> future = new CompletableFuture<>();
```

Comment:
Why do we need to reserialise on cache destroy? Keep in mind this is concurrent with cache partition operations.

Reply:
All user classes have to be removed from the cache configuration, i.e. key types, value types, all factories, etc. when tenant is destroyed, otherwise there would be class loading issues.
This is called when tenant is undeployed so there would be no other cache operations with this configuration in this case.

Reply:
As we don't have any synchronization between the proxy lifecycle and the operation lifecycle, it is still possible there are ongoing operations using tenant control, e.g. replication operations, cache read/write operations being retried because of some network issues... On the other hand, as this would be called from some unknown thread by the SPI implementation, there's not much we can do here but document that it must be invoked once there are no more ongoing operations.

Reply:
in tenant control's case, this is only called when tenant is unavailable, and in such case all operations are not going to go through (would be put back in the queue) and will eventually time out. This is the correct behavior in this instance.

Reply:
Ok, right, thanks.